### PR TITLE
Reusable proxies

### PR DESCRIPTION
**Pull request by alex-schneider**
Labeled `proxy` blocks defined in `definitions` can be referenced with the `proxy` attribute in an `endpoint`:
```hcl
server {
  api {
    endpoint "/a" {
      beta_required_permission = "pr1"
      proxy = "p"
    }
    endpoint "/a" {
      beta_required_permission = "pr2"
      proxy = "p"
    }
    endpoint "/c" {
      beta_required_permission = "pr3"
      proxy = "p"
    }
  }
}
# ...
definitions {
  proxy "p" {
    backend = "b"
    # ...
  }
  backend "b" {
    # ...
  }
}
```

---
<details>
    <summary>Reviewer checklist</summary>
    <ul>
        <li>Read PR description: a summary about the changes is required</li>
        <li>Changelog updated</li>
        <li>Documentation: docs/{Reference, Cli, ...}, Docker and cli help/usage</li>
        <li>Pulled branch, manually tested</li>
        <li>Verified requirements are met</li>
        <li>Reviewed the code</li>
        <li>Reviewed the related tests</li>
    </ul>
</details>

### CODE REVIEW

File: server/testdata/endpoints/18_couper.hcl [hunk 1]
```diff
@@ -0,0 +1,32 @@
+server "couper" {
+  endpoint "/abcdef" {
+    proxy = "test"
```
Comment (johakoch) [inline server/testdata/endpoints/18_couper.hcl:3]: This does work with free endpoints, but not with endpoints within an `api` block.

File: server/testdata/endpoints/18_couper.hcl [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    response {
```
Comment (johakoch) [inline server/testdata/endpoints/18_couper.hcl:4]: The `proxy` attribute should be a replacement for a `proxy` block, which does not require a `response` block.

Comment (alex-schneider) [reply server/testdata/endpoints/18_couper.hcl:4]: In this case its the same as:

```
endpoint "/abcdef" {
  proxy "abcdef" {
    url = "${env.COUPER_TEST_BACKEND_ADDR}/anything"
  }
}
```

The `proxy = "defaultName"` below does not require a `response` block

File: server/testdata/endpoints/18_couper.hcl (hunk 1 above)
Comment (alex-schneider) [reply server/testdata/endpoints/18_couper.hcl:3]: Fixed

File: config/configload/merge.go [hunk 3]
```diff
@@ -214,6 +214,21 @@ func mergeServers(bodies []*hclsyntax.Body, proxies map[string]*hclsyntax.Block)
 								return nil, newMergeError(errUniqueLabels, subBlock)
 							}
 
+							if attr, ok := subBlock.Body.Attributes[proxy]; ok {
```
Comment (johakoch) [inline config/configload/merge.go:217]: extract this to a function to be called for both free and api endpoints

File: server/testdata/endpoints/18_couper.hcl [hunk 4]
```diff
@@ -18,6 +18,28 @@ server "couper" {
...
+    endpoint "/api-reuse" {
+      proxy = "test"
+      response {
+        status = 204
+      }
+    }
+
+    endpoint "/api-default" {
+      proxy = "defaultName"
+      response {
```
Comment (johakoch) [inline server/testdata/endpoints/18_couper.hcl:38]: remove explicit `response` for "defaultName" proxies and expect `200`

**Review by johakoch — CHANGES_REQUESTED** [review]

File: config/configload/merge.go (hunk 3 above)
Comment (alex-schneider) [reply config/configload/merge.go]: Done

File: server/testdata/endpoints/18_couper.hcl (hunk 4 above)
Comment (alex-schneider) [reply server/testdata/endpoints/18_couper.hcl]: Done

File: config/configload/merge.go [hunk 5]
```diff
@@ -592,6 +621,34 @@ func absInBackends(block *hclsyntax.Block) error {
...
+	return nil
+}
+
+func attrStringValue(attr *hclsyntax.Attribute) (string, error) {
+	v, err := eval.Value(nil, attr.Expr)
+	if err != nil {
+		return "", err
+	}
+
+	return v.AsString(), nil
```
Comment (johakoch) [inline config/configload/merge.go:649]:
```
proxy = 1
```
Check for `cty.String` type before using `.AsString()`

File: config/configload/merge.go [hunk 6]
```diff
@@ -592,6 +621,34 @@ func absInBackends(block *hclsyntax.Block) error {
...
+	if attr, ok := block.Body.Attributes[proxy]; ok {
+		reference, err := attrStringValue(attr)
+		if err != nil {
+			return err
+		}
+
+		delete(block.Body.Attributes, proxy)
+
+		if proxyBlock, ok := proxies[reference]; !ok {
+			return newMergeError(errMissingReferencedProxy, block)
```
Comment (johakoch) [inline config/configload/merge.go:634]: * This should create an error pointing to the *attribute* (attribute's `Range`), not to the *endpoint block* (endpoint block's `DefRange`).
* Test this error

**Review by johakoch — CHANGES_REQUESTED** [review]

File: config/configload/merge.go (hunk 5 above)
Comment (johakoch) [inline config/configload/merge.go:649]:
```suggestion
	if v.Type() != cty.String {
		sr := attr.Expr.StartRange()
		return "", newDiagErr(&sr, fmt.Sprintf("%s must evaluate to string", attr.Name))
	}

	return v.AsString(), nil
```

File: config/configload/merge.go [hunk 7]
```diff
@@ -592,6 +621,34 @@ func absInBackends(block *hclsyntax.Block) error {
...
 }
 
+func addProxy(block *hclsyntax.Block, proxies map[string]*hclsyntax.Block) error {
+	if attr, ok := block.Body.Attributes[proxy]; ok {
+		reference, err := attrStringValue(attr)
+		if err != nil {
+			return err
+		}
+
+		delete(block.Body.Attributes, proxy)
```
Comment (johakoch) [inline config/configload/merge.go:631]: No need to delete this before throwing error.

File: config/configload/merge.go [hunk 8]
```diff
@@ -592,6 +621,34 @@ func absInBackends(block *hclsyntax.Block) error {
...
+
+		if proxyBlock, ok := proxies[reference]; !ok {
+			return newMergeError(errMissingReferencedProxy, block)
+		} else {
+			block.Body.Blocks = append(block.Body.Blocks, proxyBlock)
+		}
```
[the remark this answers is not in the document]
Comment (alex-schneider) [reply config/configload/merge.go:637]: No, the `proxyBlock` is only available local in `if` or `else`.

File: config/configload/merge.go (hunk 5 above)
Comment (alex-schneider) [reply config/configload/merge.go:654]: Done

File: config/configload/merge.go (hunk 6 above)
Comment (alex-schneider) [reply config/configload/merge.go:634]: Done

File: config/configload/merge.go (hunk 7 above)
Comment (alex-schneider) [reply config/configload/merge.go:631]: Done

File: config/configload/merge.go (hunk 6 above)
Comment (johakoch) [reply config/configload/merge.go]: * Test this error

is still TBD

File: config/configload/validate.go [hunk 9]
```diff
@@ -92,6 +93,11 @@ func validateBody(body hcl.Body, afterMerge bool) error {
 					if err != nil {
 						return err
 					}
+				case proxy:
+					if _, set := uniqueProxies[label]; set {
+						return newDiagErr(&labelRange, "proxy labels must be unique")
+					}
+					uniqueProxies[label] = struct{}{}
```
Comment (johakoch) [inline config/configload/validate.go:100]:
```suggestion
					if !afterMerge {
						if _, set := uniqueProxies[label]; set {
							return newDiagErr(&labelRange, "proxy labels must be unique")
						}
						uniqueProxies[label] = struct{}{}
					}
```

**Review by johakoch — APPROVED** [review]

**Pull request merged**